modules/packages/RangeChunk.chpl

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright 2004-2016 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ * 
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * 
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+module RangeChunk {
+  enum RemElems {
+    Thru,
+    Pack,
+    Mod
+  }
+  use RemElems;
+  use BoundedRangeType;
+
+  /*
+   * Range-based interface, iterator and query.
+   * idx in query is zero based.
+   */
+
+  iter chunks(
+    r: range(?RT, bounded, ?S),
+    numChunks: integral,
+    remPol: RemElems = Thru): range(RT, bounded, S)
+  {
+    for (startOrder, endOrder) in chunksOrder(r, numChunks, remPol) {
+      const start = r.orderToIndex(startOrder);
+      const end = r.orderToIndex(endOrder);
+      yield if S
+        then start..end by r.stride
+        else start..end;
+    }
+  }
+
+  proc chunk(
+    r: range(?RT, bounded, ?S),
+    numChunks: integral,
+    idx: integral,
+    remPol: RemElems = Thru): range(RT, bounded, S)
+  {
+    const (startOrder, endOrder) = chunkOrder(r, numChunks, idx, remPol);
+    const start = r.orderToIndex(startOrder);
+    const end = r.orderToIndex(endOrder);
+    return if S
+      then start..end by r.stride
+      else start..end;
+  }
+
+  /*
+   * Order-based interface, iterator and query.
+   * Orders are zero-based "indices" into a range. 
+   * idx in query is zero-based.
+   */
+
+  iter chunksOrder(
+    r: range(?RT, bounded, ?),
+    numChunks: integral,
+    remPol: RemElems = Thru): 2*RT
+  {
+    if r.length == 0 || numChunks <= 0 then
+      return;
+    const nElems = r.length;
+    const nChunks = numChunks: RT;
+
+    var chunkSize, rem: RT;
+    if remPol == Pack {
+      chunkSize = nElems / nChunks;
+      if chunkSize * nChunks != nElems then
+        chunkSize += 1;
+    } else { // Mod
+      chunkSize = nElems / nChunks;
+      rem = nElems - chunkSize * nChunks;
+    }
+
+    for i in 0..#nChunks {
+      var chunk: 2*RT;
+      if remPol == Thru {
+        chunk = chunkOrderThru(nElems, nChunks, i);
+      } else if remPol == Pack {
+         chunk = chunkOrderPack(chunkSize, nElems, i);
+      } else { // Mod
```

Comment:
I'd actually check 'if remPol == Mod' here so that when somebody adds a new enum constant it doesn't just fall into this case. Same for the other places that have an if/then/else looking for all of the enum values.

Or better, a switch statement that accepts all of the known enum values and issues an error in the 'otherwise' case.